Implement table view on Web UI

196Ikuchil opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
- Reorganization of the directories of some current resources List components.
- Implement `DataTable` on web UI. This table view is switchable to the current resources list view.
  - This is an alpha feature. It will be enabled if the `ALPHA_TABLE_VIEWS` environment value is `1`.
- simple searching.
- sorting.
- clickable to edit and delete.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #96

#### Special notes for your reviewer:
##### Samples
<img width="700" alt="スクリーンショット 2022-02-28 22 57 05" src="https://user-images.githubusercontent.com/22634362/155995536-5e77e067-02f1-4fa9-b9f9-b7ea3151c7d5.png">
<img width="700" alt="スクリーンショット 2022-02-28 22 57 16" src="https://user-images.githubusercontent.com/22634362/155995546-9876ac30-f6b1-4939-9c94-c8ede5a9b9fa.png">
<img width="700" alt="スクリーンショット 2022-02-28 23 06 15" src="https://user-images.githubusercontent.com/22634362/155996724-c2fb7e1b-19f4-4079-9fce-93aa8785d40c.png">



##### clickable
<img width="700" alt="スクリーンショット 2022-02-28 22 58 12" src="https://user-images.githubusercontent.com/22634362/155995638-8c1f9e62-0300-475d-92a9-c27cd49cc722.png">




/label tide/merge-method-squash

196Ikuchil commented on web/components/ResourceViews/DataTables/NodeDataTable.vue:
```diff
@@ -0,0 +1,86 @@
+      nstore.select(node, false);
+    };
+    const search = "";
+    const headers = [
```
# Need help
We need to discuss which columns should be selected for display.
I think that displaying all parameters is not good.
The current columns of each data table are my randomly selected.

196Ikuchil replied:
In commit [d225fbc](https://github.com/kubernetes-sigs/kube-scheduler-simulator/pull/119/commits/d225fbcfdd17243f58278aff835786695d698215), I changed some columns(headers).

These columns are displayed in each table.
- name
- namespace
- capacity
- status(phase)
- condition
- creation-time
- update-time

And for other columns, the unique values of each resource were chosen.

sanposhiho commented:
/ok-to-test

196Ikuchil commented on web/components/ResourceViews/DataTables/NodeDataTable.vue:
```diff
@@ -0,0 +1,100 @@
+      },
+      {
+        text: "Update-Time",
+        value: "metadata.managedFields[0].time",
```
currently, we display the `managed field`'s timestamp as an update time.

196Ikuchil replied:
I think that there is no way to get the updated time except this.
I also checked that this doesn't raise an error even if managedFields[0] is nil. (The cell becomes empty.)

sanposhiho commented:
/assign

sanposhiho commented:
@196Ikuchil

Some feedbacks. 

For Pod table, 
- `Status` is not needed. (Because it's always pending).

For Node table, 
- `Namespace` is not needed. Node is not namespaced resource. 
- `Conditions ` and `Status` is not needed.
- `Capacity` should show each resource(CPU, memory, and pod)

For PV table, 
- `Namespace` is also not needed. not namespaced resource. 
-  `PVReclaimPolicy` is not needed.

For PVC table, 
- `Status` is also not needed. same reason as Pod.

For SC table, 
- `Namespace` is not needed. same reason.
- `Parameters`, `Reclaim-Policy`, and `Volume-Binding-Mode` are not needed.

For PC table,
- `Namespace` is not needed. same reason.

And for all tables, 
- `Update-Time` is not needed. (Sorry I said that it was needed tho 🙏 )
- The length of the Search box should be the same for all tables.
- change not to use `-`. Use the exact name instead. (like CreationTime, GlobalDefault...)

Others
- Move "List, Table" switch to top bar.

sanposhiho commented:
And, in current implementation, `Node` on `Pod` table is not updated.
I raise it as another issue because this is not a problem specific to this feature.
https://github.com/kubernetes-sigs/kube-scheduler-simulator/issues/120

196Ikuchil commented:
Hi @sanposhiho 
Thanks to the reviews. I generally understand.
I have two questions.

1.  `Status `
On `PVC`, there are possibilities to take other values. Is `status` not needed?[Ref:kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#persistentvolumeclaimstatus-v1-core)
(Note: On other resources, I have checked that the `status` parameter is Deprecated. For example, [PodStatus](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#podstatus-v1-core). Therefore, I will remove it from the table.)

2. Move "List, Table" switch to top bar.
I think it shouldn't move. This is because the buttons should not be placed far from the table. Isn't it hard to tell what the switch is for? 
And also the button's component file should be placed under the `ResourceViews` directory.

sanposhiho commented:
> On PVC, there are possibilities to take other values. Is status not needed?

Yes. Not needed. This value is not much related to scheduling. (This is "scheduler" simulator, so columns that are not related to scheduling are not needed basically.)

> This is because the buttons should not be placed far from the table. Isn't it hard to tell what the switch is for?

It's a fair point. But it is the same even in its current location. It is only when the user touches it that they realize it has something to do with how it is displayed.

How about removing the button and change the UI with env.
Like `ALPHA__TABLE_VIEWS`. If `ALPHA__TABLE_VIEWS` == true (or 1?), use the table to show resources.

sanposhiho commented:
I think one of the strengths of the current view is "It's easy to see where Pod is."

Conversely, if we can achieve this strength in table view, we can replace the current UI with table view completely.

sanposhiho commented:
/retitle Implement table view on Web UI

sanposhiho commented:
> Conversely, if we can achieve this strength in this table view, I think we can replace the current UI with table view completely.

This does not mean that I want you to make changes that way in this PR. I think we can merge this PR first, and then make improvements for that. (Otherwise, this PR will become too big.)

196Ikuchil commented:
> How about removing the button and change the UI with env.
Like ALPHA__TABLE_VIEWS. If ALPHA__TABLE_VIEWS == true (or 1?), use the table to show resources.

Understood. In this PR, we use `env` value. 
And we expect that the current view will replace this table view in nearly future.

196Ikuchil commented:
@sanposhiho 
Fixed it✋
I also updated the documentation and screenshots at the top of this PR.
https://github.com/kubernetes-sigs/kube-scheduler-simulator/pull/119#issue-1148249561

196Ikuchil commented:
/unhold

sanposhiho commented on web/components/ResourceViews/DataTables/DataTable.vue:
```diff
@@ -0,0 +1,59 @@
+    };
+  },
+});
+</script>
```
```suggestion
</script>

<style>
.row-pointer > .v-data-table__wrapper > table > tbody > tr:hover {
  cursor: pointer;
}
</style>

```

sanposhiho commented on web/components/ResourceViews/DataTables/DataTable.vue:
```diff
@@ -0,0 +1,59 @@
+          :items="items"
+          :items-per-page="5"
+          :search="search"
+          multi-sort
```
```suggestion
          multi-sort
          class="row-pointer"
```

sanposhiho reviewed (approved):
Thanks @196Ikuchil 👍 
/lgtm
/approve
